web-api/src/mind.rs

```diff
@@ -41,6 +45,39 @@ struct Article {
     url: String,
 }
 
+#[derive(Debug, Deserialize)]
+struct ArticleProvider {
+    articles: HashMap<String, Article>,
+}
+
+impl ArticleProvider {
+    fn new(path: &str) -> Result<Self, anyhow::Error> {
+        let mut articles = HashMap::new();
+        for article in read::<Article>(path)? {
+            let article = article?;
+            articles.insert(article.id.clone(), article);
+        }
+        Ok(Self { articles })
+    }
+
+    fn sample(&self, n: usize) -> Vec<Article> {
```

Comment:
We could think about directly sorting out the retrieved elements from the dataset. However, we could also wait for further logic to know what exactly we want to achieve and how we will reset the dataset for every user